File: src/main/java/com/lessmarkup/userinterface/nodehandlers/configuration/ConfigurationRootNodeHandler.java
package com.lessmarkup.userinterface.nodehandlers.configuration;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import com.google.inject.Inject;
import com.lessmarkup.Constants;
import com.lessmarkup.TextIds;
import com.lessmarkup.framework.helpers.DependencyResolver;
import com.lessmarkup.framework.helpers.LanguageHelper;
import com.lessmarkup.framework.helpers.StringHelper;
import com.lessmarkup.framework.nodehandlers.AbstractNodeHandler;
import com.lessmarkup.interfaces.cache.DataCache;
import com.lessmarkup.interfaces.module.ModuleConfiguration;
import com.lessmarkup.interfaces.module.ModuleProvider;
import com.lessmarkup.interfaces.structure.*;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.OptionalLong;

class ConfigurationGroupData {
    private final String title;
    private final List<ConfigurationHandlerData> handlers = new ArrayList<>();
    
    public ConfigurationGroupData(String title) {
        this.title = title;
    }
    
    public String getTitle() {
        return this.title;
    }
    
    public List<ConfigurationHandlerData> getHandlers() {
        return this.handlers;
    }
}

class ConfigurationHandlerData {
    private Class<? extends NodeHandler> type;
    private String titleTextId;
    private String title;
    private String moduleType;
    private long id;
    private String typeName;
    
    public Class<? extends NodeHandler> getType() {
        return this.type;
    }
    
    public void setType(Class<? extends NodeHandler> value) {
        this.type = value;
    }
    
    public String getTitleTextId() {
        return this.titleTextId;
    }
    
    public void setTitleTextId(String value) {
        this.titleTextId = value;
    }
    
    public String getTitle() {
        return this.title;
    }
    
    public void setTitle(String value) {
        this.title = value;
    }
    
    public String getModuleType() {
        return this.moduleType;
    }
    
    public void setModuleType(String value) {
        this.moduleType = value;
    }
    
    public long getId() {
        return this.id;
    }
    
    public void setId(long value) {
        this.id = value;
    }
    
    public String getTypeName() {
        return this.typeName;
    }
    
    public void setTypeName(String value) {
        this.typeName = value;
    }
}

public class ConfigurationRootNodeHandler extends AbstractNodeHandler {

    private final ModuleProvider moduleProvider;
    private final DataCache dataCache;
    private final List<ConfigurationGroupData> configurationGroups = new ArrayList<>();
    private final HashMap<String, ConfigurationHandlerData> configurationHandlers = new HashMap<>();

    @Inject
    public ConfigurationRootNodeHandler(ModuleProvider moduleProvider, DataCache dataCache) {
        this.moduleProvider = moduleProvider;
        this.dataCache = dataCache;
        
        addScript("scripts/controllers/ConfigurationController");
        
        long idCounter = 1;
        
        ConfigurationGroupData normalGroup = new ConfigurationGroupData(LanguageHelper.getFullTextId(Constants.ModuleType.MAIN, TextIds.SITE_CONFIGURATION));
        
        for (ModuleConfiguration module : this.moduleProvider.getModules()) {
            for (Class<? extends NodeHandler> nodeHandlerType : module.getInitializer().getNodeHandlerTypes()) {
                ConfigurationHandler configurationHandler = nodeHandlerType.getAnnotation(ConfigurationHandler.class);
                if (configurationHandler == null) {
                    continue;
                }
                String typeName = nodeHandlerType.getSimpleName().toLowerCase();
                if (typeName.endsWith("nodehandler")) {
                    typeName = typeName.substring(0, typeName.length()-"nodehandler".length());
                }
                
                ConfigurationHandlerData handlerData = new ConfigurationHandlerData();
                handlerData.setType(nodeHandlerType);
                handlerData.setModuleType(module.getModuleType());
                handlerData.setTitleTextId(configurationHandler.titleTextId());
                handlerData.setId(idCounter++);
                handlerData.setTypeName(typeName);
                
                this.configurationHandlers.put(typeName, handlerData);
                normalGroup.getHandlers().add(handlerData);
            }
        }
        
        if (!normalGroup.getHandlers().isEmpty()) {
            this.configurationGroups.add(normalGroup);
        }
        
        for (ConfigurationHandlerData handler : this.configurationHandlers.values()) {
            handler.setTitle(LanguageHelper.getFullTextId(handler.getModuleType(), handler.getTitleTextId()));
        }
        
        for (ConfigurationGroupData group : this.configurationGroups) {
            group.getHandlers().sort((h1, h2) -> h1.getTitle().compareTo(h2.getTitle()));
        }
    }
    
    @Override
    public boolean hasChildren() {
        return true;
    }
    
    @Override
    public JsonObject getViewData() {
        String path = getObjectId().isPresent() ? this.dataCache.get(NodeCache.class).getNode(getObjectId().getAsLong()).getFullPath() : null;
        
        JsonObject ret = new JsonObject();
        
        JsonArray groups = new JsonArray();
        
        for (ConfigurationGroupData g : this.configurationGroups) {
            JsonObject group = new JsonObject();
            group.addProperty("title", g.getTitle());
            JsonArray items = new JsonArray();
            g.getHandlers().forEach(h -> {
                JsonObject handler = new JsonObject();
                handler.addProperty("path", path != null ? (path + "/" + h.getTypeName()) : h.getTypeName());
                handler.addProperty("title", LanguageHelper.getFullTextId(h.getModuleType(), h.getTitleTextId()));
                items.add(handler);
            });
            group.add("items", items);
            groups.add(group);
        }
        
        ret.add("groups", groups);
        
        return ret;
    }
    
    @Override
    public ChildHandlerSettings getChildHandler(String path) {
        List<String> parts = new ArrayList<>();
        Arrays.stream(path.split("/")).map(p -> p.trim()).forEach(parts::add);
        if (parts.isEmpty()) {
            return null;
        }
        
        ConfigurationHandlerData handlerData = this.configurationHandlers.get(parts.get(0));
        if (handlerData == null) {
            return null;
        }
        
        NodeHandler handler = DependencyResolver.resolve(handlerData.getType());
        
        path = parts.get(0);
        parts.remove(0);
        
        handler.initialize(OptionalLong.empty(), null, path, getFullPath() + "/" + path, NodeAccessType.WRITE);
        
        ChildHandlerSettings ret = new ChildHandlerSettings();
        
        ret.setId(OptionalLong.of(handlerData.getId()));
        ret.setHandler(handler);
        ret.setTitle(LanguageHelper.getText(handlerData.getModuleType(), handlerData.getTitleTextId()));
        ret.setPath(path);
        ret.setRest(StringHelper.join(",", parts));
        
        return ret;
    }
}

File: src/main/scala/com/lessmarkup/userinterface/nodehandlers/configuration/ConfigurationRootNodeHandler.scala
/*
 * This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
 * If a copy of the MPL was not distributed with this file, You can obtain one at
 * http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.userinterface.nodehandlers.configuration

import com.google.gson.{JsonArray, JsonObject}
import com.google.inject.Inject
import com.lessmarkup.framework.helpers.LanguageHelper
import com.lessmarkup.framework.nodehandlers.{AbstractNodeHandler, NodeHandlerConfiguration}
import com.lessmarkup.interfaces.annotations._
import com.lessmarkup.interfaces.cache.DataCache
import com.lessmarkup.interfaces.module.ModuleProvider
import com.lessmarkup.interfaces.structure._

class ConfigurationGroupData(val title: String, val handlers: Seq[ConfigurationHandlerData])

class ConfigurationHandlerData(
  val handlerType: Class[_ <: NodeHandlerFactory],
  val titleTextId: String,
  val moduleType: String,
  val groupTextId: String,
  val title: String,
  val id: Long,
  val typeName: String)

class ConfigurationRootNodeHandlerFactory @Inject() (moduleProvider: ModuleProvider, dataCache: DataCache) extends NodeHandlerFactory {
  override def createNodeHandler(nodeHandlerConfiguration: NodeHandlerConfiguration, arguments: Any*): NodeHandler = {
    new ConfigurationRootNodeHandler(moduleProvider, dataCache, nodeHandlerConfiguration)
  }
}

class ConfigurationRootNodeHandler(moduleProvider: ModuleProvider, dataCache: DataCache, configuration: NodeHandlerConfiguration)
  extends AbstractNodeHandler(configuration) {

  private val allHandlers: Seq[ConfigurationHandlerData] = {
    (for (
      ((module, handlerType), index) <- moduleProvider.getModules
        .flatMap(m => m.getInitializer.getNodeHandlerTypes.map(t => (m, t)))
        .iterator.zipWithIndex;
      configurationHandler = handlerType.getAnnotation(classOf[ConfigurationHandler])
      if configurationHandler != null
    ) yield {

        val fullTypeName = handlerType.getSimpleName.toLowerCase
        val minimizedTypeName =
          if (fullTypeName.endsWith("nodehandler"))
            fullTypeName.substring(0, fullTypeName.length - "nodehandler".length)
          else
            fullTypeName

        new ConfigurationHandlerData(
          handlerType = handlerType,
          moduleType = module.getModuleType,
          titleTextId = configurationHandler.titleTextId,
          title = LanguageHelper.getFullTextId(module.getModuleType, configurationHandler.titleTextId),
          id = index,
          typeName = minimizedTypeName,
          groupTextId = configurationHandler.groupTextId()
        )
      }).toSeq
  }

  private val configurationGroups: Seq[ConfigurationGroupData] = allHandlers.groupBy(_.groupTextId).map {
      case (groupTextId, handlers) =>
        new ConfigurationGroupData(groupTextId, handlers.sortBy(_.title))
    }.toSeq

  private final val configurationHandlers: Map[String, ConfigurationHandlerData] = allHandlers.map(h => (h.typeName, h)).toMap

  override def getScripts: Seq[String] = super.getScripts :+ "scripts/controllers/ConfigurationController"

  override def hasChildren: Boolean = true

  override def getViewData: Option[JsonObject] = {
    val path: String = if (getObjectId.isDefined) dataCache.get(classOf[NodeCache]).getNode(getObjectId.get).get.fullPath else null
    val ret: JsonObject = new JsonObject
    val groups: JsonArray = new JsonArray
    for (g <- this.configurationGroups) {
      val group: JsonObject = new JsonObject
      group.addProperty("title", g.title)
      val items: JsonArray = new JsonArray
      for (h <- g.handlers) {
        val handler = new JsonObject
        handler.addProperty("path", if (path != null) path + "/" + h.typeName else h.typeName)
        handler.addProperty("title", LanguageHelper.getFullTextId(h.moduleType, h.titleTextId))
        items.add(handler)
      }

      group.add("items", items)
      groups.add(group)
    }
    ret.add("groups", groups)
    Option(ret)
  }

  override def createChildHandler(path: Seq[String]): Option[ChildHandlerSettings] = {

    if (path.isEmpty) {
      return None
    }

    val handlerData = configurationHandlers.get(path.head)
    if (handlerData.isEmpty) {
      return None
    }

    val nodeHandlerConfiguration = new NodeHandlerConfiguration(
      objectId = Option(handlerData.get.id),
      settings = None,
      accessType = NodeAccessType.WRITE,
      path = path.head,
      fullPath = getFullPath + "/" + path.head
    )

    val handler = createChildHandler(handlerData.get.handlerType, nodeHandlerConfiguration)

    Option(new ChildHandlerSettings(
      handler = handler,
      id = nodeHandlerConfiguration.objectId,
      title = handlerData.get.title,
      path = path.head,
      rest = path.tail
    ))
  }
}